ast/env.go

```diff
@@ -304,6 +304,116 @@ func (n *typeTreeNode) Put(path Ref, tpe types.Type) {
 	curr.value = tpe
 }
 
+// Insert inserts tpe at path in the tree, but also merges the value into any types.Object present along that path.
+// If an types.Object is inserted, any leafs already present further down the tree are merged into the inserted object.
+// path must be ground.
+func (n *typeTreeNode) Insert(path Ref, tpe types.Type) {
+	curr := n
+	for i, term := range path {
+		c, ok := curr.children.Get(term.Value)
+
+		var child *typeTreeNode
+		if !ok {
+			child = newTypeTree()
+			child.key = term.Value
+			curr.children.Put(child.key, child)
+		} else {
+			child = c.(*typeTreeNode)
+
+			if child.value != nil && i+1 < len(path) {
+				// If child hase an object value, merge the new value into it.
+				if o, ok := child.value.(*types.Object); ok {
+					var err error
+					child.value, err = insertIntoObject(o, path[i+1:], tpe)
+					if err != nil {
+						panic(fmt.Errorf("unreachable, insertIntoObject: %w", err))
+					}
+				}
```

Comment:
Could the `!ok` case happen? If so, what happens to our loop, what do we set `curr` to?

Reply:
`curr` would be set to `child`, the next node in the tree along the `path`. So we'd blissfully just keep moving further down the tree to update it and ignore that the node had any value at all. However, this scenario would imply that there is a declared rule at e.g. `data.a.b` that defines some non-object value, and we're currently updating the type for e.g. some rule `data.a.b.c`. The only time the compiler will allow one rule to be within the extent of another is if the outer most rule (the one with the shortest path) is an object.
This behavior is in line with `Put()`, which does not care what value any of the existing tree nodes have when a type is added. Which I think is sensible at this point  as we have knowledge about some of the restrictions posed by the compiler, and can opt-out of unnecessary work.